File: src/ua/edu/hneu/oop/java/lr5/task2/Menu.java
package ua.edu.hneu.oop.java.lr5.task2;

import ua.edu.hneu.oop.java.util.ConsoleUtil;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Menu {
    private static final String INVITATION = "Enter a command:%n>>>";
    private static final String HELP_DESCRIPTION = "Print this help.";
    private static final String DESCRIPTION_FORMAT = "%-10s  ---  %s";
    private static final String PATTERN_FORMAT = "(?i)^%s";
    private static final String[] HELP_COMMAND_KEYS = {"h", "help"};
    private static final String COMMAND_NOT_FOUND = "Entered command not found. For help enter help.";
    private List<String> menuDescriptions;
    private Map<String, VoidMethod> menu;
    private ConsoleUtil consoleUtil;

    public Menu(ConsoleUtil consoleUtil) {
        this.consoleUtil = consoleUtil;
        menuDescriptions = new ArrayList<>();
        menu = new HashMap<>();
        addMenuRow(HELP_DESCRIPTION, this::printHelp, HELP_COMMAND_KEYS);
    }

    public void addMenuRow(String description, VoidMethod action, String... keys) {
        appendDescription(description, keys);
        menu.put(toPattern(keys), action);
    }

    public void process() throws IOException, ClassNotFoundException {
        consoleUtil.printf(INVITATION);
        processInput(consoleUtil.readInput());
    }

    public void printHelp() {
        menuDescriptions.forEach(s -> consoleUtil.println(s));
        consoleUtil.printf("%n");
    }

    private void processInput(String command) throws IOException, ClassNotFoundException {
        menu.entrySet()
                .stream()
                .filter(p -> command.matches(p.getKey()))
                .map(Map.Entry::getValue)
                .findAny()
                .orElse(this::defaultCommand).call();
    }

    private void defaultCommand() {
        consoleUtil.println(COMMAND_NOT_FOUND);
    }

    private void appendDescription(String description, String... key) {
        menuDescriptions.add(String.format(DESCRIPTION_FORMAT, keysJoining(key), description));
    }

    private String toPattern(String... key) {
        return String.format(PATTERN_FORMAT, keysJoining(key));
    }

    private String keysJoining(String... key) {
        return String.join("|", key);
    }
}

File: src/ua/edu/hneu/oop/dotNet/LR5/Task2/Menu.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Task2
{
    public class Menu
    {
        private const String Invitation = "Enter a command:\n>>>";
        private const String HelpDescription = "Print this help.";
        private const String DescriptionFormat = "{0,-10}  ---  {1}";
        private const String PatternFormat = "{0}";
        private const String CommandNotFound = "Entered command not found. For help enter help.";

        private static readonly String[] HelpCommandKeys = {"h", "help"};

        private readonly List<String> _menuDescriptions = new List<string>();
        private readonly Dictionary<String, Action> _menuItems = new Dictionary<string, Action>();

        public Menu()
        {
            AddMenuRow(HelpDescription, PrintHelp, HelpCommandKeys);
        }

        public void AddMenuRow(String description, Action action, params string[] keys)
        {
            AppendDescription(description, keys);
            _menuItems.Add(ToPattern(keys), action);
        }

        public void Process()
        {
            Console.WriteLine(Invitation);
            ProcessInput(Console.ReadLine());
        }

        public void PrintHelp()
        {
            foreach (string description in _menuDescriptions)
                Console.WriteLine(description);
            Console.WriteLine("%n");
        }

        private void ProcessInput(String command)
        {
            string key = _menuItems.Keys.FirstOrDefault(x => x.Split('|').Contains(command));
            if (key == null)
                DefaultCommand();
            else
                _menuItems[key]();
        }

        private void DefaultCommand() => Console.WriteLine(CommandNotFound);

        private void AppendDescription(String description, params string[] keys)
        {
            _menuDescriptions.Add(String.Format(DescriptionFormat, KeysJoining(keys), description));
        }

        private String ToPattern(params string[] keys) => String.Format(PatternFormat, KeysJoining(keys));

        private String KeysJoining(params string[] keys) => String.Join("|", keys);
    }
}